### PR TITLE
Added Shell Sort [CPP]

### LINKED ISSUE #192: Shell Sort [CPP]
-   https://en.wikipedia.org/wiki/Shellsort

### PR DESCRIPTION
**Pull request by mohitkyadav**
Fixes #192 

#### By submitting this pull request I confirm I've read and complied with the below declarations.

- [x] I have read the [Contribution guidelines](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md) and I am confident that my PR reflects them.
- [x] I have followed the [coding guidelines](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md#cs) for this project.
- [x] My code follows the [skeleton code structure](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md#sample).
- [x] This pull request has a descriptive title. For example, `Added {Algorithm/DS name} [{Language}]`, not `Update README.md` or `Added new code`.
- [x] This pull request will be closed if I fail to update it even once in a continuous timespan of 7 days.

### CODE REVIEW

File: shell_sort/ShellSort.cpp [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+#include <cmath>
+
+using namespace std;
+
+// Worst case time complexity = O(n^2)
+// Best case complexity = O(nlog(n))
+
+vector<int> ShellSort(vector<int> data) {
+    for (int i = floor(data.size() / 2); i > 0; i /= 2) {
+        for (int j = i; j < floor(data.size()); j++) {
```
Comment (singhpratyush) [inline shell_sort/ShellSort.cpp]: `floor`is redundant.

File: shell_sort/ShellSort.cpp [hunk 2]
```diff
@@ -0,0 +1,38 @@
+#include <iostream>
+#include <vector>
+#include <cmath>
+
+using namespace std;
+
+// Worst case time complexity = O(n^2)
+// Best case complexity = O(nlog(n))
+
+vector<int> ShellSort(vector<int> data) {
```
Comment (singhpratyush) [inline shell_sort/ShellSort.cpp]: Please follow either `lowerCamelCase` or `under_scores`.

**Comment by singhpratyush** [comment]
Also, can you mention somewhere to use `-std=c++11` flag while compiling? It can be confusing at times.

File: shell_sort/ShellSort.cpp [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+// g++ -std=c++11 -o ShellSort ShellSort.cpp
+
+using namespace std;
+
+// Worst case time complexity = O(n^2)
+// Best case complexity = O(nlog(n))
+
+vector<int> shellSort(vector<int> data) {
+	for (int i = static_cast<int>(data.size() / 2); i > 0; i /= 2) {
+		for (int j = i; j < data.size(); j++) {
```
Comment (Monal5031) [inline shell_sort/ShellSort.cpp]: `++j` is better than `j++` I think :P

Comment (mohitkyadav) [inline shell_sort/ShellSort.cpp]: I also think... 😄 ...  👍

File: shell_sort/ShellSort.cpp [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+int main() {
+	vector<int> data = {1000, 45, -45, 121, 47, 45, 65, 121, -1, 103, 45, 34};
+	cout << "Data to be sorted:" << '\n';
+	for (auto item : data) {
+		cout << item << " ";
+	}
+	cout << '\n' << "Sorted data:" << '\n';
+	data = shellSort(data);
+	for (auto item : data) {
```
Comment (singhpratyush) [inline shell_sort/ShellSort.cpp]: DRY. Please use a function instead.

File: shell_sort/ShellSort.cpp [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+// Use '-std=c++11' or higher flag while compiling
+// g++ -std=c++11 -o ShellSort ShellSort.cpp
+
+using namespace std;
+
+// Worst case time complexity = O(n^2)
+// Best case complexity = O(nlog(n))
+
+vector<int> shellSort(vector<int> data) {
```
Comment (Monal5031) [inline shell_sort/ShellSort.cpp]: Shouldn't function name in c++ be separated using underscores rather than using `camelCase`.

Comment (mohitkyadav) [inline shell_sort/ShellSort.cpp]: > Please follow either lowerCamelCase or under_scores.

**Comment by aviaryan** [comment]
Comment on 8910d341149a4b6b3df031057455bd8746d38411, file modular_exponential/modular_exponential.py, line 3.

The code does not comply to PEP8.

*PEP8Bear, severity NORMAL, section `pep8`.*

The issue can be fixed by applying the following patch:

```diff
--- a/modular_exponential/modular_exponential.py
+++ b/modular_exponential/modular_exponential.py
@@ -1,6 +1,7 @@
 """
 Modular Exponential
 """
+
 
 def mod_exponent(base, power, mod):
     """
```

**Comment by aviaryan** [comment]
Comment on 8910d341149a4b6b3df031057455bd8746d38411, file stack/stack.py, line 1.

The code does not comply to PEP8.

*PEP8Bear, severity NORMAL, section `pep8`.*

The issue can be fixed by applying the following patch:

```diff
--- a/stack/stack.py
+++ b/stack/stack.py
@@ -1,4 +1,5 @@
 class Stack(object):
+
     def __init__(self):
         self._list = []
 
```

**Comment by mohitkyadav** [comment]
@aviaryan should I fix errors in `modular_exponential.py` and `stack.py`.

**Comment by Monal5031** [comment]
@mohitkyadav No need I have created a separate issue for it, also sent a PR for it. 😃

**Comment by Monal5031** [comment]
@mohitkyadav Please rebase and squash your commits to pass all the tests

**Comment by mohitkyadav** [comment]
I've already rebased my local repo. Now both the repositories are even and there are no conflicts. Since it's merged into master there's no need.

**Comment by Monal5031** [comment]
@mohitkyadav Sorry somehow I missed that rebased part, was using github on mobile. :smiley:

**Pull request merged**